exporter/signalfxexporter/translation/translator.go

```diff
@@ -327,6 +347,54 @@ func (mp *MetricTranslator) TranslateDataPoints(logger *zap.Logger, sfxDataPoint
 					convertMetricValue(logger, dp, newType)
 				}
 			}
+		case ActionCalculateNewMetric:
+			var operand1, operand2 *sfxpb.DataPoint
+			for _, dp := range processedDataPoints {
+				if dp.Metric == tr.Operand1Metric {
+					operand1 = dp
+				} else if dp.Metric == tr.Operand2Metric {
+					operand2 = dp
+				}
+			}
+
+			if operand1 == nil {
+				logger.Warn("calculate_new_metric: operand1 == nil")
+				continue
+			}
+			if operand1.Value.IntValue == nil {
+				logger.Warn("calculate_new_metric: operand1.Value.IntValue == nil")
+				continue
+			}
+
+			if operand2 == nil {
+				logger.Warn("calculate_new_metric: operand2 == nil")
+				continue
+			}
+			if operand2.Value.IntValue == nil {
+				logger.Warn("calculate_new_metric: operand2.Value.IntValue == nil")
+				continue
+			}
+
+			if tr.Operator == MetricOperatorDivision && *operand2.Value.IntValue == 0 {
+				logger.Warn("tr.Operator == MetricOperatorDivision && *operand2.Value.IntValue == 0")
+				continue
+			}
+
+			newPt := proto.Clone(operand1).(*sfxpb.DataPoint)
+			newPt.Metric = tr.MetricName
+			var newPtVal float64
+			switch tr.Operator {
+			// only supporting divide operator for now
+			case MetricOperatorDivision:
+				// only supporting int values for now
+				newPtVal = float64(*operand1.Value.IntValue) / float64(*operand2.Value.IntValue)
+			default:
+				logger.Warn("calculate_new_metric: unsupported operator", zap.String("operator", string(tr.Operator)))
+				continue
+			}
+			newPt.Value = sfxpb.Datum{DoubleValue: &newPtVal}
```

Comment:
This case block seems to be pretty big. What do you think of moving this part to a separate function?

Reply:
Agree and fixed. I wonder if there's a better way longer term so that this code abides by the open-closed principle. Like, instead of this switch, we pass in all of the supported rules at runtime, and when a new one is needed, we create it separately and add it to the list of rules that gets passed in.

Reply:
Sounds good. we can improve it later